\documentclass{article}
\usepackage[sc]{mathpazo}
\renewcommand{\sfdefault}{lmss}
\renewcommand{\ttdefault}{lmtt}
\usepackage[T1]{fontenc}
\usepackage{geometry}
\geometry{verbose,tmargin=2.5cm,bmargin=2.5cm,lmargin=2.5cm,rmargin=2.5cm}
\setcounter{secnumdepth}{2}
\setcounter{tocdepth}{2}
\usepackage[unicode=true,pdfusetitle,
 bookmarks=true,bookmarksnumbered=true,bookmarksopen=true,bookmarksopenlevel=2,
 breaklinks=false,pdfborder={0 0 1},backref=false,colorlinks=false]
 {hyperref}
\hypersetup{
 pdfstartview={XYZ null null 1}}
\usepackage{breakurl}

\makeatletter
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% User specified LaTeX commands.
\renewcommand{\textfraction}{0.05}
\renewcommand{\topfraction}{0.8}
\renewcommand{\bottomfraction}{0.8}
\renewcommand{\floatpagefraction}{0.75}

\makeatother

\begin{document}

<<setup,include=FALSE,cache=FALSE>>=
  opts_chunk$set(fig.width=8, fig.height=6, fig.path = 'figure')
@

% \section{General Information}
% File: \Sexpr{x@filename}\\
% Type: \Sexpr{getType(x)}\\
% Sequence Length Range: \Sexpr{seqLengthRange(x)[1]} to \Sexpr{seqLengthRange(x)[2]}\\
% Total Sequences: \Sexpr{sum(x@seq.lengths)}\\
% Sequences hashed: \Sexpr{x@hashed}\\
% Unique Sequences: \Sexpr{if (!(ll <- length(x@hash))) NA else ll}\\
% k-mers hashed: \Sexpr{x@kmers.hashed}\\
% k: \Sexpr{if (!length(x@k)) NA else x@k}\\
% Proportion reads hashed: \Sexpr{if (!length(x@hash.prop)) NA else x@hash.prop}\\
  


\section{Quality by Position}
<<qualplot, echo=FALSE, message=FALSE, fig.align="left">>=
  if (getType(x) == "FASTQ") {
    qualPlot(x)
  } else {
    cat("No quality data for FASTA file.")
  }
@ 

\section{Nucleotide Proportion by Position}
<<baseplot, echo=FALSE, message=FALSE, fig.align="left">>=
  basePlot(x, type="proportion")
@ 

  
The R session information (including the OS info, R version and all
packages used):
<<session-info, cache=FALSE>>=
sessionInfo()
Sys.time()
@
<<clean-up, include=FALSE>>=
rm(.knitr.author); rm(.knitr.author)
@

\end{document}
